packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -63,8 +63,62 @@ async function findAllHotswappableChanges(
   let foundNonHotswappableChange = false;
   const promises: Array<Array<Promise<ChangeHotswapResult>>> = [];
 
+  // as a first pass, we need to collapse replacement changes into one,
+  // as they are represented here as a pair of two changes: one addition and one removal
+  const allResourceChanges: { [logId: string]: cfn_diff.ResourceDifference } = stackChanges.resources.changes;
```

Comment:
This function was a right beast to review. Some extracting out of code to helper functions would have helped a lot.

Like

```ts
function filterDict<A>(xs: Record<string, A>, predicate: (x: A) => boolean): Record<string, A>;

function sameResource(a: ResourceDifference, b: ResourceDifference): boolean;

function makeRenameDifference(removed: ResourceDifference, added: ResourceDifference): ResourceDifference;
```

...or something

Reply:
Agreed 100%. Implemented all 3.